### ISSUE TITLE
Gnome 45: import declarations may only appear at top level of a module

### ISSUE THREAD

**Issue by guilhem** [ddterm/gnome-shell-extension-ddterm#527]
### GNOME Shell version

45.beta.1

### Session type

Wayland (`echo $WAYLAND_DISPLAY` output is non-empty)

### Special configuration

- [X] High DPI
- [X] More than one monitor

### ddterm version

44

### Linux distro

Ubuntu Mantic Minotaur (development branch)

### Description

Error:
```
SyntaxError: import declarations may only appear at top level of a module

Stack trace:
  @file:///home/guilhem/.local/share/gnome-shell/extensions/ddterm@amezin.github.com/prefs.js:24:12
  _init/GLib.MainLoop.prototype.runAsync/</<@resource:///org/gnome/gjs/modules/core/overrides/GLib.js:266:34
```

This looks to be related to https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1984#note_1275832

**Comment by guilhem** [comment]
https://gjs.guide/extensions/upgrading/gnome-shell-45.html#esm

**Comment by stouracp** [comment]
Any news about gome 45 supported version?

**Comment by amezin** [comment]
The issue here is that as soon as you make the extension compatible with v45, it's no longer compatible with any previous version.

I'll try to make the extension compatible with v45 around September, 20 (GNOME 45 release date, Fedora 39 beta, Ubuntu 23.10 Beta). But before that I need to make one release for previous GNOME versions (hopefully this week).

**Comment by guilhem** [comment]
The recommend way is to have 2 branches with and without esm import.

https://blogs.gnome.org/shell-dev/2023/09/02/extensions-in-gnome-45/

Gnome extension manage to have multiple versions of the same extension.

**Comment by stouracp** [comment]
Yes, give us a gnome 45 branch, and we will hapily test it.

**Comment by amezin** [comment]
> The recommend way is to have 2 branches with and without esm import.

Which means the same as "maintain 2 separate extensions". And I'm currently unable to keep up with 1.

Since compatibility breaks anyway, I see this as an opportunity to upgrade to newer JS syntax, switch to Gtk 4, maybe libadwaita. But all of this will make backporting hard if not impossible.

So the idea is: complete the features I've been working on (split layout, D-Bus API, maybe session save/restore), make a release, then switch completely to GNOME >= 45.

**Comment by stouracp** [comment]
So, gnome 45 is released, when we can have ddterm for it? Some testing branch?

**Comment by garfunkel** [comment]
After updating to 45 and finding my own (small) extension break (again) and having to fix it (again) for the new Gnome version, I completely understand how horrible developing and testing extensions truly are. Salute to the folk who develop ddterm.

The Gnome recommendation to simply maintain two branches is quite outrageous.

**Comment by horga83** [comment]
I'm not entirely sure I understand this.  I do agree that things are a bit of a pain but if you make a GIT branch for 45 from your master which would presumably be 44 and 44 is working, then you pretty much only have to maintain the 45 branch, so back to one again.  I suppose the odd bug maybe but really it's only one branch moving forward.

Am I missing something?  Anyway thanks for your work on ddterm, I do really enjoy using it.

**Comment by amezin** [comment]
https://github.com/ddterm/gnome-shell-extension-ddterm/tree/gnome45

**Comment by guilhem** [comment]
> https://github.com/ddterm/gnome-shell-extension-ddterm/tree/gnome45

Thanks @amezin it works <3

Only problem I see (but I can open another bug) terminal doesn't show off with shortcut (mine is "œ" (^2) when I'm focused on the terminal (it works when focused outside)
![image](https://github.com/ddterm/gnome-shell-extension-ddterm/assets/486876/2f86d212-135e-470e-8645-53395add9017)

**Comment by amezin** [comment]
No, please don't open bugs about GNOME 45 until it's in `master` branch.

> terminal doesn't show off with shortcut (mine is "œ" (^2) when I'm focused on the terminal (it works when focused outside)

I don't understand, could you rephrase? Especially what you mean by "[doesn't] show off"

**Comment by guilhem** [comment]
It's doesn't "toggle off" (I don't exactly the term).

If I press "œ" → term open
If I press again "œ" → character "œ" appears in terminal.
If I change focus (click elsewhere), press "œ" → term show off.

**Comment by amezin** [comment]
Hm. ddterm disables the keybinding only when you disable the extension.

Do you test it with ddterm being the only window on screen? (i. e. either ddterm window is focused, or desktop) What happens if there's also some other window - file manager, for example?

Could you try changing the shortcut to `F12`, or try adding `<Ctrl>`?

**Comment by guilhem** [comment]
> Hm. ddterm disables the keybinding only when you disable the extension.
> 
> Do you test it with ddterm being the only window on screen? (i. e. either ddterm window is focused, or desktop) What happens if there's also some other window - file manager, for example?
> 
> Could you try changing the shortcut to `F12`, or try adding `<Ctrl>`?

Yes it did work with F12 and ctrl+²
That's too bad because ² was the “quake console” key :)

**Comment by stouracp** [comment]
> https://github.com/ddterm/gnome-shell-extension-ddterm/tree/gnome45

That works. Awesome, thank you a lot.

**Comment by amezin** [comment]
https://github.com/ddterm/gnome-shell-extension-ddterm/releases/tag/v48
